design/ironic_authentication.md

```diff
@@ -0,0 +1,329 @@
+* Start a reverse proxy locally (in a container) that uses the client
+  certificate to connect to inspector and ironic
+* Configure clouds.yaml to connect without authentication to the local reverse
+  proxy
+* Start the current openstack cli container with the command arguments, that
+  will then connect to the local reverse proxy that will authenticate towards
+  the ironic and inspector web servers.
+
+The setup would look like the following (red is authenticated, green is unauthenticated)
+
+![mTLS setup](images/ironic_authentication_mTLS.png)
+
+#### Impact on pivoting
+
+In order to perform the pivoting, a second instance of CAPM3 and BMO is started
+on the target cluster. A second instance of ironic cannot however be started
+until the first instance is removed, to not duplicate the DHCP. Since ironic is
+using keepalived to share the IP address, this means that the second BMO
+instance, deployed with a new secret, will try to communicate with the first
+ironic instance and fail, since the CA that signed the first ironic instance
+certificate is unknown to the second BMO instance. There would also be a client
+certificate mismatch in addition to the server certificate mismatch. This would
+be resolved when the first ironic instance is removed and the second ironic
+instance deployed, finalizing the pivoting.
+
+It is possible to overcome the issue by using a CA provided by the user.
+
+### Risks and Mitigations
+
+A Keystone based solution is too complex for the added benefits (considering
+that BMO is mostly the only client of Ironic and inspector), while the basic
+authentication might not be sufficiently secured for some of our stakeholders.
+
+Adding a web server with a wsgi solution also increases the complexity of the
+setup. However, it is proven to be a successful production setup in multiple
+ironic setups.
+
+Using mTLS adds a high level of complexity to use any other clients than BMO.
+For example to use the Openstack CLI we would need a solution such as a local
+reverse proxy taking care of the authentication. However, the CLI is only used
+for debugging scenarios in development environments, hence having a lower
+impact.
+
+Running BMO locally would then require having a local copy of the client
+certificate, or having an unencrypted and unauthenticated setup option. An
+insecure environment can be kept for development purposes.
+
+### Work Items
+
+This is to implement the mTLS solution.
+
+* Create certificates properly with cert-manager (also for client if mTLS)
+* Run Ironic and inspector using a web server and a wsgi solution
+* Configure the web server to use those certificates, and offer unencrypted
+  endpoints on the loopback interface.
+* Modify BMO to accept the server certificate (and authenticate with the
+  client certificate if mTLS), using the certificates form the secrets
+* Modify the Openstack CLI script to run a local proxy if needed and
+  authenticate.
+
+### Dependencies
+
+None
```

Comment:
Should we capture the changes required to Gophercloud to make this work? Does it work out of the box with the NoAuth auth mechanism we have in there already?

Reply:
Disregard, I see the discussion of custom http clients under Design.